cmd/booster-bitswap/client.go

```diff
@@ -0,0 +1,220 @@
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/p2p/muxer/mplex"
+	"github.com/libp2p/go-libp2p/p2p/muxer/yamux"
+	quic "github.com/libp2p/go-libp2p/p2p/transport/quic"
+	"github.com/libp2p/go-libp2p/p2p/transport/tcp"
+	"github.com/urfave/cli/v2"
+	"golang.org/x/sync/errgroup"
+)
+
+var fetchCmd = &cli.Command{
+	Name:        "fetch",
+	Usage:       "fetch <multiaddr> <root cid> <output car path>",
+	Description: "Fetch all blocks in the DAG under the given root cid from the bitswap node at multiaddr",
+	Before:      before,
+	Flags: []cli.Flag{
+		&cli.BoolFlag{
+			Name:  "pprof",
+			Usage: "run pprof web server on localhost:6071",
+		},
+		&cli.IntFlag{
+			Name:  "concurrency",
+			Usage: "concurrent request limit - 0 means unlimited",
+			Value: 10,
+		},
+	},
+	Action: func(cctx *cli.Context) error {
+		if cctx.Bool("pprof") {
+			go func() {
+				err := http.ListenAndServe("localhost:6071", nil)
+				if err != nil {
+					log.Error(err)
+				}
+			}()
+		}
+
+		if cctx.Args().Len() != 3 {
+			return fmt.Errorf("usage: fetch <multiaddr> <root cid> <output car path>")
+		}
+
+		addrInfoStr := cctx.Args().Get(0)
+		serverAddrInfo, err := peer.AddrInfoFromString(addrInfoStr)
+		if err != nil {
+			return fmt.Errorf("parsing server multiaddr %s: %w", addrInfoStr, err)
+		}
+
+		rootCidStr := cctx.Args().Get(1)
+		rootCid, err := cid.Parse(rootCidStr)
+		if err != nil {
+			return fmt.Errorf("parsing cid %s: %w", rootCidStr, err)
+		}
+
+		outputCarPath := cctx.Args().Get(2)
+
+		ctx := lcli.ReqContext(cctx)
+
+		// setup libp2p host
+		log.Infow("generating libp2p key")
+		privKey, _, err := crypto.GenerateECDSAKeyPair(rand.Reader)
+		if err != nil {
+			return err
+		}
+
+		host, err := libp2p.New(
+			libp2p.Transport(tcp.NewTCPTransport),
+			libp2p.Transport(quic.NewTransport),
+			libp2p.Muxer("/mplex/6.7.0", mplex.DefaultTransport),
+			libp2p.Muxer("/yamux/1.0.0", yamux.DefaultTransport),
+			libp2p.Identity(privKey),
+			libp2p.ResourceManager(network.NullResourceManager),
+		)
+		if err != nil {
+			return err
+		}
+
+		// Create a bitswap client
+		nilRouter, err := nilrouting.ConstructNilRouting(ctx, nil, nil, nil)
+		if err != nil {
+			return err
+		}
+		net := bsnetwork.NewFromIpfsHost(host, nilRouter)
+		bs, err := blockstore.OpenReadWrite(outputCarPath, []cid.Cid{rootCid}, blockstore.UseWholeCIDs(true))
+		if err != nil {
+			return fmt.Errorf("creating blockstore at %s: %w", outputCarPath, err)
+		}
+
+		ctx, cancel := context.WithCancel(ctx)
+		defer cancel()
+		brn := &blockReceiver{bs: bs, ctx: ctx, cancel: cancel}
+		bsClient := client.New(ctx, net, bs, client.WithBlockReceivedNotifier(brn))
+		defer bsClient.Close()
+		net.Start(bsClient)
+
+		// Connect to host
+		connectStart := time.Now()
+		log.Infow("connecting to server", "server", serverAddrInfo.String())
+		err = host.Connect(ctx, *serverAddrInfo)
+		if err != nil {
+			return fmt.Errorf("connecting to %s: %w", serverAddrInfo, err)
```

Comment:
This will likely cause you problems with large files since if the connection gets pruned on the server side it'll never get re-established and your download will stall (e.g. like https://github.com/ipfs/ipget/issues/103).

It's resolvable by trying to keep the connection alive.